src/main/java/hudson/plugins/git/Cache.java

```diff
@@ -0,0 +1,102 @@
+package hudson.plugins.git;
+
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.Node;
+
+import jenkins.model.Jenkins;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+public class Cache {
+
+    /**
+     * Keep one lock per cache directory. Lazy populated, but never purge, except on restart.
+     */
+    private static final ConcurrentMap<String, ConcurrentMap<String, Lock>> cacheLocks = new ConcurrentHashMap<>();
+
+    private static final Logger LOGGER = Logger.getLogger(Cache.class.getName());
+
+    private static String getCacheEntry(String remoteURL) {
+        return "git-" + Util.getDigestOf(remoteURL);
+    }
+
+    private static String getNodeEntry(Node node) {
+        return Util.getDigestOf(node.getNodeName().isEmpty() ? node.getNodeName() : "master");
+    }
+
+    public static FilePath getCacheDir(Node node, String remoteURL)
+            throws IOException, InterruptedException {
+        String cacheEntry = getCacheEntry(remoteURL);
+        FilePath cacheDir = new FilePath(new FilePath(node.getRootPath(), "caches"), cacheEntry);
+        cacheDir.mkdirs(); // ensure it exists
+        return cacheDir;
+    }
+
+    public static File getCacheDir(String remoteURL) {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins == null) {
+            LOGGER.severe("Jenkins instance is null in Cache.getCacheDir");
+            return null;
+        }
+        File cacheDir = new File(new File(jenkins.getRootDir(), "caches"), getCacheEntry(remoteURL));
+        if (!cacheDir.isDirectory()) {
+            boolean ok = cacheDir.mkdirs();
+            if (!ok) {
+                LOGGER.log(Level.WARNING, "Failed mkdirs of {0}", cacheDir);
+            }
+        }
+        return cacheDir;
+    }
+
+    // Get node lock
+    private static Lock getCacheLock(Node node, String cacheEntry) {
+        ConcurrentMap<String, Lock> cacheNodeLocks;
+        Lock cacheLock;
+        String nodeEntry = getNodeEntry(node);
+        while (null == (cacheNodeLocks = cacheLocks.get(nodeEntry))) {
+            cacheLocks.putIfAbsent(nodeEntry, new ConcurrentHashMap<String, Lock>());
+        }
+        while (null == (cacheLock = cacheNodeLocks.get(cacheEntry))) {
+            cacheNodeLocks.putIfAbsent(cacheEntry, new ReentrantLock());
+        }
+        return cacheLock;
+    }
+
+    // Lock/unlock master
+    public static void lock(String remoteURL) {
+        Jenkins jenkins = Jenkins.getInstance();
+        if (jenkins != null) {
+            lock(jenkins, remoteURL);
+        } else {
+            LOGGER.severe("Jenkins instance is null in Cache.lock");
```

Comment:
@oleg-nenashev this method has been moved from AbstractGitSCMSource untouched: [link](https://github.com/jenkinsci/git-plugin/blob/master/src/main/java/jenkins/plugins/git/AbstractGitSCMSource.java#L670). If I throw an exception from there, I'll have two failing tests (unexpected exception):
- AbstractGitSCMSourceRetrieveHeadsTest.correctGitToolIsUsed
- AbstractGitSCMSourceRetrieveHeadsTest.correctGitToolIsUsed2
I cannot suppress cache use for these tests because it uses master cache in GitSCMFileSystem. I'm not sure if it's a test issue, and don't see the right way to fix it. Or should  I fix it in scope of this change at all.